optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ListChangeMove.java

```diff
@@ -0,0 +1,223 @@
+     *                   └ sourceEntity
+     * }
+     * </pre>
+     *
+     * <h4>Example 1 - source and destination entities are different</h4>
+     *
+     * <pre>
+     * {@code
+     * GIVEN
+     * Ann.tasks = [A, B, C]
+     * Bob.tasks = [X, Y]
+     *
+     * WHEN
+     * ListChangeMove: A {Ann[0]->Bob[2]}
+     *
+     * THEN
+     * Ann.tasks = [B, C]
+     * Bob.tasks = [X, Y, A]
+     * }
+     * </pre>
+     *
+     * <h4>Example 2 - source and destination is the same entity</h4>
+     *
+     * <pre>
+     * {@code
+     * GIVEN
+     * Ann.tasks = [A, B, C]
+     *
+     * WHEN
+     * ListChangeMove: A {Ann[0]->Ann[2]}
+     *
+     * THEN
+     * Ann.tasks = [B, C, A]
+     * }
+     * </pre>
+     *
+     * @param variableDescriptor descriptor of a list variable, for example {@code Employee.taskList}
+     * @param sourceEntity planning entity instance from which a planning value will be removed, for example "Ann"
+     * @param sourceIndex index in sourceEntity's list variable from which a planning value will be removed
+     * @param destinationEntity planning entity instance to which a planning value will be moved, for example "Bob"
+     * @param destinationIndex index in destinationEntity's list variable where the moved planning value will be inserted
+     */
+    public ListChangeMove(
+            ListVariableDescriptor<Solution_> variableDescriptor,
+            Object sourceEntity, int sourceIndex,
+            Object destinationEntity, int destinationIndex) {
+        this.variableDescriptor = variableDescriptor;
+        this.sourceEntity = sourceEntity;
+        this.sourceIndex = sourceIndex;
+        this.destinationEntity = destinationEntity;
+        this.destinationIndex = destinationIndex;
+    }
+
+    public Object getSourceEntity() {
+        return sourceEntity;
+    }
+
+    public int getSourceIndex() {
+        return sourceIndex;
+    }
+
+    public Object getDestinationEntity() {
+        return destinationEntity;
+    }
+
+    public int getDestinationIndex() {
+        return destinationIndex;
+    }
+
+    public Object getMovedValue() {
+        return variableDescriptor.getElement(sourceEntity, sourceIndex);
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public boolean isMoveDoable(ScoreDirector<Solution_> scoreDirector) {
+        // TODO maybe remove this because no such move should be generated
+        // Do not use Object#equals on user-provided domain objects. Relying on user's implementation of Object#equals
+        // opens the opportunity to shoot themselves in the foot if different entities can be equal.
+        return destinationEntity != sourceEntity
+                || (destinationIndex != sourceIndex && destinationIndex != variableDescriptor.getListSize(sourceEntity));
```

Comment:
Can the index be actually bigger than the list size?
```suggestion
                || (destinationIndex != sourceIndex && destinationIndex < variableDescriptor.getListSize(sourceEntity));
```

Reply:
> Can the index be actually bigger than the list size?

`MoveSelector` never creates such moves. But the user can create arbitrary moves and execute them. We do that in `SolutionBusiness`:
```java
move.doMoveOnly(guiScoreDirector);
```

The way I see it, `isMoveDoable()` is only used to filter out moves that we can't avoid generating in the `MoveSelector` either because their doability cannot be determined at that time or because skipping them in the `MoveSelector` would be to difficult (complicated code, esp. with random selection).

So  in the interest of performance I'd like to avoid an in-depth parameter validation:
```java
if (sourceIndex < 0 || destinationIndex < 0) {
    // MoveSelector never creates such ListMove, but users can...
    return false;
}
return destinationEntity != sourceEntity && destinationIndex <= variableDescriptor.getListSize(destinationEntity)
        || (destinationIndex != sourceIndex && destinationIndex < variableDescriptor.getListSize(sourceEntity));
```

If, on the other hand, defensive code is the priority, then we should do that. Wdyt @ge0ffrey?